### PR TITLE
Condition Logic Checker

### PR DESCRIPTION
**Pull request by Raine-Yang-UofT**
## Proposed Changes

_(Describe your changes here. Also describe the motivation for your changes: what problem do they solve, or how do they improve the application or codebase? If this pull request fixes an open issue, [use a keyword to link this pull request to the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).)_

Implement a custom checker for the following messages:
- `redundant-condition`: a conditional statement is guaranteed true given the control flow leading to the statement.
- `impossible-condition`: a conditional statement is guaranteed false either due to contradition within the statement or contradiction with conditions leading to the statement.

These two messages are enabled only if Z3 dependency is installed and `z3` option is turned on for PythonTA

Additional changes:
In `check-examples`, the `z3` option is set to `True` to check for `redundant-condition` and `impossible-condition`

...

<details>
<summary>Screenshots of your changes (if applicable)</summary>

</details>

## Type of Change

_(Write an `X` or a brief description next to the type or types that best describe your changes.)_

| Type                                                                                    | Applies? |
| --------------------------------------------------------------------------------------- | -------- |
| 🚨 _Breaking change_ (fix or feature that would cause existing functionality to change) |          |
| ✨ _New feature_ (non-breaking change that adds functionality)                          |     X     |
| 🐛 _Bug fix_ (non-breaking change that fixes an issue)                                  |          |
| ♻️ _Refactoring_ (internal change to codebase, without changing functionality)          |          |
| 🚦 _Test update_ (change that _only_ adds or modifies tests)                            |          |
| 📚 _Documentation update_ (change that _only_ updates documentation)                    |          |
| 📦 _Dependency update_ (change that updates a dependency)                               |          |
| 🔧 _Internal_ (change that _only_ affects developers or continuous integration)         |          |

## Checklist

_(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)_

Before opening your pull request:

- [x] I have performed a self-review of my changes.
  - Check that all changed files included in this pull request are intentional changes.
  - Check that all changes are relevant to the purpose of this pull request, as described above.
- [x] I have added tests for my changes, if applicable.
  - This is **required** for all bug fixes and new features.
- [x] I have updated the project documentation, if applicable.
  - This is **required** for new features.
- [x] I have updated the project Changelog (this is required for all changes).
- [x] If this is my first contribution, I have added myself to the list of contributors.

After opening your pull request:

- [x] I have verified that the pre-commit.ci checks have passed.
- [x] I have verified that the CI tests have passed.
- [x] I have reviewed the test coverage changes reported by Coveralls.
- [x] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments

_(Include any questions or comments you have regarding your changes.)_

### CODE REVIEW

File: examples/custom_checkers/r9900_redundant_condition.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+def return_large_number(x: int) -> int:
+    """
+    Return number x only if it's greater than 1000
```
Comment (david-yz-liu) [inline examples/custom_checkers/r9900_redundant_condition.py:3]: The description should always start on the same line as the triple-quotes.

File: examples/custom_checkers/r9900_redundant_condition.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+    Preconditions:
+        - x > 1000
+    """
+    if x > 1000:
```
Comment (david-yz-liu) [inline examples/custom_checkers/r9900_redundant_condition.py:8]: Leave a comment indicating the error is on this line, and for all examples.

Follow the style in the existing files.

File: examples/custom_checkers/r9900_redundant_condition.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+        return x
+
+
+def nested_condition(x: int) -> int:
+    if x > 10:
+        if x > 5:
```
Comment (david-yz-liu) [inline examples/custom_checkers/r9900_redundant_condition.py:14]: same as above

File: examples/custom_checkers/r9901_impossible_condition.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
+def print_natural_number(x: int):
+    """
+    Print number x only if it's a natural number.
```
Comment (david-yz-liu) [inline examples/custom_checkers/r9901_impossible_condition.py:3]: See my above comment about descriptions.

But also pick a different example, don't introduce the terminology of natural numbers here (this is unnecessarily complex for these examples)

File: examples/custom_checkers/r9901_impossible_condition.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+    Preconditions:
+        - x >= 0
+    """
+    if x < 0:
+        raise Exception("x is not a natural number")
+    print(x)
+
+
+def invalid_condition(x: bool):
```
Comment (david-yz-liu) [inline examples/custom_checkers/r9901_impossible_condition.py:14]: Don't use the term "invalid" here; "impossible" is fine.

Also, this example is quite good. The parallel example for the redundant condition checker would be `x or not x`; please add this example to that file.

File: python_ta/checkers/redundant_or_impossible_condition_checker.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    z3 = Any
+    z3_dependency_available = False
+
+from astroid import nodes
+from pylint.checkers import BaseChecker
+from pylint.checkers.utils import only_required_for_messages
+from pylint.lint import PyLinter
+
+
+class RedundantOrImpossibleConditionChecker(BaseChecker):
```
Comment (david-yz-liu) [inline python_ta/checkers/redundant_or_impossible_condition_checker.py:21]: Rename this class (and corresponding file, etc.) to `ConditionLogicChecker`

File: python_ta/checkers/redundant_or_impossible_condition_checker.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    name = "redundant-condition"
+    msgs = {
+        "R9900": (
+            """This conditional statement is already true given the preceding code;
```
Comment (david-yz-liu) [inline python_ta/checkers/redundant_or_impossible_condition_checker.py:25]: Reword to "This condition will always evaluate to True." Similarly for the `impossible-condition` error.

Note: As one of your examples above demonstrates, this checker will flag conditions that are tautological independent of the surrounding code, e.g. `x or not x`. So it's misleading to reference "preceding code" or, below, "preceding Z3 constraints".

File: python_ta/checkers/redundant_or_impossible_condition_checker.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+             consider removing redundant check.""",
+            "redundant-condition",
+            "Used when an If or While statement is tautologically true based on preceding Z3 constraints."
+            " Enabled when Z3 option is on.",
```
Comment (david-yz-liu) [inline python_ta/checkers/redundant_or_impossible_condition_checker.py:29]: Enabling and disabling is a configuration option, not automatic. But a good thing to note here is "Requires the z3 configuration option to be True."

File: python_ta/checkers/redundant_or_impossible_condition_checker.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+            {
+                "default": False,
+                "type": "yn",
+                "metavar": "<y or n>",
+                "help": "Use Z3 to perform logical feasibility analysis in program control flow.",
+            },
+        ),
+    )
+
+    def __init__(self, linter=None) -> None:
```
Comment (david-yz-liu) [inline python_ta/checkers/redundant_or_impossible_condition_checker.py:49]: This isn't necessary to define, since it's just calling the superclass initializer

File: python_ta/checkers/redundant_or_impossible_condition_checker.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        super().__init__(linter=linter)
+
+    @only_required_for_messages("redundant-condition", "impossible-condition")
+    def visit_if(self, node: nodes.If) -> None:
+        """Visit if statement"""
+        self._check_condition(node)
+        # handle else branch
```
Comment (david-yz-liu) [inline python_ta/checkers/redundant_or_impossible_condition_checker.py:56]: delete this comment

File: python_ta/checkers/redundant_or_impossible_condition_checker.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+        """Check if the condition is impossible."""
+        solver = z3.Solver()
+        solver.add(prev_constraints)
+        solver.add(node_constraint)
+        return solver.check() == z3.unsat
+
+
+def register(linter: PyLinter) -> None:
+    """Required method to auto-register this checker to the linter,
+    Register the linter only if `z3` option is turned on.
```
Comment (david-yz-liu) [inline python_ta/checkers/redundant_or_impossible_condition_checker.py:115]: Delete this sentence (and make sure to end the preceding clause with a period, not a comma).

File: python_ta/checkers/redundant_or_impossible_condition_checker.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+            return
+
+        node_block = node.cfg_block
+
+        if not node_block.is_feasible or node_block.z3_constraint is None:
+            return
+
+        for pred in node_block.predecessors:
+            if all(
+                self._check_redundant_condition(
```
Comment (david-yz-liu) [inline python_ta/checkers/redundant_or_impossible_condition_checker.py:81]: Overall, you don't need to define separate functions for the two different conditions. You just need one function, `_check_unsat`, which is given a list of `z3.ExprRef`s and returns whether the `And` of those is unsatisfiable. You can use such a function both here and below. For the redundant-condition, check whether the constraints plus the _negation_ of the `node_block`'s constraint is unsatisfiable.

File: python_ta/checkers/redundant_or_impossible_condition_checker.py [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+                    z3.And(*[c for c in constraints]), node_block.z3_constraint
```
Comment (david-yz-liu) [inline python_ta/checkers/redundant_or_impossible_condition_checker.py:82]: `constraints` is itself an iterable, so you can just do `*constraints`

File: python_ta/checkers/redundant_or_impossible_condition_checker.py [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+            or not self.linter.config.z3
+        ):
+            return
+
+        node_block = node.cfg_block
+
+        if not node_block.is_feasible or node_block.z3_constraint is None:
+            return
+
+        for pred in node_block.predecessors:
```
Comment (david-yz-liu) [inline python_ta/checkers/redundant_or_impossible_condition_checker.py:79]: only include feasible edges here; this also removes the need for a `node_block.is_feasible` check above, as if there are no feasible predecessor edges the loop won't run at all

File: python_ta/cfg/graph.py [hunk 15]
```diff
@@ -286,29 +286,31 @@ def update_edge_z3_constraints(self) -> None:
...
-                    if condition_z3_constraint is not None:
-                        if edge.negate is not None:
-                            z3_environment.add_constraint(
-                                Not(condition_z3_constraint)
-                                if edge.negate
-                                else condition_z3_constraint
-                            )
-
-                edge.z3_constraints[path_id] = z3_environment.update_constraints()
+                    if condition_z3_constraint is not None and edge.negate is not None:
```
Comment (david-yz-liu) [inline python_ta/cfg/graph.py:289]: Overall, you should not perform unnecessary refactoring. Revert the changes in this file, and see my comments below.

File: python_ta/checkers/redundant_or_impossible_condition_checker.py [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+        if (
+            not hasattr(node, "cfg_block")
+            or not z3_dependency_available
+            or not self.linter.config.z3
+        ):
+            return
+
+        node_block = node.cfg_block
+
+        if not node_block.is_feasible or node_block.z3_constraint is None:
```
Comment (david-yz-liu) [inline python_ta/checkers/redundant_or_impossible_condition_checker.py:76]: Rather than create a separate instance attribute `z3_constraint`:

1. Extract the `condition` of the `node` object
2. Create a new `Z3Environment` from the variables from the underlying CFG (you can define a new `ControlFlowGraph` to produce this environment).
3. Use the environment to parse the `condition` expression directly, in this function.

Comment (Raine-Yang-UofT) [reply python_ta/checkers/redundant_or_impossible_condition_checker.py]: I'm still working on this part and am unsure how to proceed. The main issue is that I cannot determine how to access the `ControlFlowGraph` instance associated with a node. Currently, I can only access the node's `CFGBlock`, which does not contain a reference back to the `ControlFlowGraph`. 

My previous approach of adding `z3_constraint` to nodes could be the most convenient implementation. I understand your design decision to avoid embedding too much information directly into the control flow graph itself. However, even if I find a way to access the `ControlFlowGraph`, we would still need a mechanism to track the reassignment status of variables within the current block, as this is not currently handled by the CFG.

In summary, I’m currently unable to parse the condition effectively without modifying the `graph.py` module. I’d appreciate any suggestions you might have on how to proceed without altering the core module or if adjustments to it might be acceptable.

Comment (david-yz-liu) [reply python_ta/checkers/redundant_or_impossible_condition_checker.py]: @Raine-Yang-UofT this is a good summary of your current understanding. You said:

> The main issue is that I cannot determine how to access the `ControlFlowGraph` instance associated with a node. Currently, I can only access the node's `CFGBlock`, which does not contain a reference back to the `ControlFlowGraph`.

Every astroid node can access its enclosing *frame* (either a function definition or the `Module`), I think the method is called `.frame()` or something similar. From this you can access the enclosing CFG.

Comment (Raine-Yang-UofT) [reply python_ta/checkers/redundant_or_impossible_condition_checker.py]: @david-yz-liu Hi Professor, thanks for this hint. However, there is another tricky problem even after accessing `ControlFlowGraph` - knowing the reassignment status of each variables.

In our implementation, we discard the variables being reassigned in subsequent z3 constraints. As a result, to correctly parse the node condition, we need not only the z3 variables provided by `ControlFlowGraph`, but also which variables are  reassigned when we reach the condition node, information we do not have even with `ControlFlowGraph`.

In addition, as the variable reassignment status may be different along different execution paths, it is possible that on different execution paths the node condition constraints should be parsed differently. 

Given these problems, the most convenient solution might be adding `z3_constraints` to `CFGBlock` during graph traverse, where we have all the information above. Let me know how should I proceed.

Comment (david-yz-liu) [reply python_ta/checkers/redundant_or_impossible_condition_checker.py]: Hi @Raine-Yang-UofT, whether or not variables are reassigned should impact the constraints stored in the edges, but this should *not* impact the conversion of the if/while condition.

The algorithm used by these checkers should always use the translated if/while condition. The reassignments will impact how this condition is combined with the existing path constraints that reach the condition.

File: CHANGELOG.md [hunk 17]
```diff
@@ -43,6 +43,8 @@ and adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 ### 💫 New checkers
 
 - `unmentioned-parameter`: Provide error message when a function parameter is not mentioned by name in the function's docstring. By default, this checker is disabled.
+- `redundant-condition`: Provide error message when a conditional statement within a function is guaranteed true. This checker requires `z3` option to be turned on.
```
Comment (david-yz-liu) [inline CHANGELOG.md:46]: These should be moved into the `unreleased` section (note that I made a version update recently, so this part of the changelog now refers to 2.9.0).

File: docs/checkers/index.md [hunk 18]
```diff
@@ -2699,6 +2699,48 @@ python_ta.check_all(config={
...
+
+This error occurs when an `if` or `while` condition is guaranteed to be true.
+
+Example:
+
+```{literalinclude} /../examples/custom_checkers/r9900_redundant_condition.py
+
+```
+
+This error will only be checked if `z3` dependency is install and `z3` option of PythonTA
```
Comment (david-yz-liu) [inline docs/checkers/index.md:2714]: `z3` dependency -> the `z3-solver` library

install -> installed

is turned on -> is enabled

File: docs/checkers/index.md [hunk 19]
```diff
@@ -2699,6 +2699,48 @@ python_ta.check_all(config={
...
+
+This error occurs when an `if` or `while` condition is guaranteed to be false.
+
+Example:
+
+```{literalinclude} /../examples/custom_checkers/r9901_impossible_condition.py
+
+```
+
+This error will only be checked if `z3` dependency is install and `z3` option of PythonTA
```
Comment (david-yz-liu) [inline docs/checkers/index.md:2734]: similar to above

File: examples/custom_checkers/r9900_redundant_condition.py [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+
+def nested_condition(x: int) -> int:
+    if x > 10:
+        # the condition `x > 5` is already guaranteed by `x > 10`
+        if x > 5:   # Error on this line
+            return x
+    return 0
+
+
+def redundant_condition(x: bool):
```
Comment (david-yz-liu) [inline examples/custom_checkers/r9900_redundant_condition.py:20]: include a return type annotation of `None` here

File: examples/custom_checkers/r9901_impossible_condition.py [hunk 21]
```diff
@@ -0,0 +1,28 @@
+def print_none_negative_number(x: int):
```
Comment (david-yz-liu) [inline examples/custom_checkers/r9901_impossible_condition.py:1]: include a return type annotation

File: examples/custom_checkers/r9901_impossible_condition.py [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+    Preconditions:
+        - x >= 0
+    """
+    # the if condition is impossible given the function precondition
+    if x < 0:   # Error on this line
+        raise Exception("x is smaller than zero")
+    print(x)
+
+
+def impossible_condition(x: bool):
```
Comment (david-yz-liu) [inline examples/custom_checkers/r9901_impossible_condition.py:13]: include a return type here

File: examples/custom_checkers/r9901_impossible_condition.py [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+    # the if condition is always false
+    if x and not x:     # Error on this line
+        print("impossible")
+
+
+def Display_number(x: int) -> str:
```
Comment (david-yz-liu) [inline examples/custom_checkers/r9901_impossible_condition.py:19]: name this function with a lowercase `display`

File: python_ta/checkers/condition_logic_checker.py [hunk 24]
```diff
@@ -0,0 +1,102 @@
+"""
+Check for redundant If or While conditions in functions based on z3 constraints
```
Comment (david-yz-liu) [inline python_ta/checkers/condition_logic_checker.py:2]: redundant -> redundant/impossible

File: python_ta/checkers/condition_logic_checker.py [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+from pylint.checkers.utils import only_required_for_messages
+from pylint.lint import PyLinter
+
+
+class ConditionLogicChecker(BaseChecker):
+    name = "redundant-condition"
+    msgs = {
+        "R9900": (
+            """This condition will always evaluate to True;
+             consider removing redundant check.""",
```
Comment (david-yz-liu) [inline python_ta/checkers/condition_logic_checker.py:29]: Remove this second part and end the sentence after `True`. ("consider" does not convey the right level of instruction here, as we really do want students to remove the redundant condition)

File: python_ta/checkers/condition_logic_checker.py [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+        self._check_condition(node)
+
+    @only_required_for_messages("redundant-condition", "impossible-condition")
+    def visit_while(self, node: nodes.While) -> None:
+        """Visit while statement"""
+        self._check_condition(node)
+
+    def _check_condition(self, node: Union[nodes.If, nodes.While]) -> None:
+        """A condition statement is redundant if for every feasible execution path
+        leading to the node, the condition must be True by precedent constraints.
```
Comment (david-yz-liu) [inline python_ta/checkers/condition_logic_checker.py:63]: This explanation is good, but you should make the following two changes:

1. The first line of the docstring should indicate what the function does, not provide a definition. Something like `Check whether `node.test` is redundant or impossible.`
2. You should include a parallel definition for "impossible" condition, since this is also relevant for this method.

File: python_ta/checkers/condition_logic_checker.py [hunk 27]
```diff
@@ -0,0 +1,102 @@
...
+
+        if all(
+            self._check_unsat(z3.And(*constraints), z3_condition)
+            for edge in (pred for pred in node_block.predecessors if pred.is_feasible)
+            for constraints in edge.z3_constraints.values()
+        ):
+            self.add_message("impossible-condition", node=node)
+
+    def _check_unsat(self, prev_constraints: z3.ExprRef, node_constraint: z3.ExprRef) -> bool:
+        """Check if the condition is redundant."""
```
Comment (david-yz-liu) [inline python_ta/checkers/condition_logic_checker.py:94]: This docstring isn't quite right. This function returns whether the `And` of its two inputs is unsatisfiable.

File: tests/test_examples.py [hunk 28]
```diff
@@ -149,7 +150,7 @@ def test_examples_files_pyta(test_file: str, pyta_examples_symbols: dict[str, se
     found_pylint_message = checker_name in file_symbols
     assert (
         found_pylint_message
-    ), f"Failed {test_file}. File does not add expected message  {file_symbols}."
+    ), f"Failed {test_file}. File does not add expected message {checker_name}:  {file_symbols}."
```
Comment (david-yz-liu) [inline tests/test_examples.py:153]: This change is okay, but please delete the extra space after the colon

File: python_ta/checkers/condition_logic_checker.py [hunk 29]
```diff
@@ -0,0 +1,102 @@
...
+        condition_node = node.test
+        env = Z3Environment(node.frame().cfg.z3_vars, [])
+        z3_condition = env.parse_constraint(condition_node)
+
+        if all(
+            self._check_unsat(z3.And(*constraints), z3.Not(z3_condition))
+            for edge in (pred for pred in node_block.predecessors if pred.is_feasible)
+            for constraints in edge.z3_constraints.values()
+        ):
+            self.add_message("redundant-condition", node=node)
```
Comment (david-yz-liu) [inline python_ta/checkers/condition_logic_checker.py:84]: pass `node.test` as the argument (so that only the condition is highlighted)

File: python_ta/checkers/condition_logic_checker.py [hunk 30]
```diff
@@ -0,0 +1,102 @@
...
+
+        if all(
+            self._check_unsat(z3.And(*constraints), z3_condition)
+            for edge in (pred for pred in node_block.predecessors if pred.is_feasible)
+            for constraints in edge.z3_constraints.values()
+        ):
+            self.add_message("impossible-condition", node=node)
```
Comment (david-yz-liu) [inline python_ta/checkers/condition_logic_checker.py:91]: same as above

File: python_ta/checkers/condition_logic_checker.py [hunk 31]
```diff
@@ -0,0 +1,102 @@
+"""
+Check for redundant If or While conditions in functions based on z3 constraints
+"""
+
```
Comment (david-yz-liu) [inline python_ta/checkers/condition_logic_checker.py:4]: Include a `from __future__ import annotations`, otherwise the type annotation `z3.ExprRef` below will raise an error if `z3` is not installed

**Review by david-yz-liu — COMMENTED** [review]
@Raine-Yang-UofT great work! I left a few inline comments, which should be pretty small.

I did note one oddity: both the redundant and impossible condition checks were being triggered on my top-level `if __name__ == '__main__'` condition, even when they shouldn't have been. I'm not sure exactly why this is, can you please look into it?

**Review by david-yz-liu — APPROVED** [review]
Thank you for all of your work on this, @Raine-Yang-UofT!
